src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2018 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.spark.sql.v2
+
+import com.mongodb.MongoClient
+import com.mongodb.client.model.{InsertOneModel, ReplaceOneModel, ReplaceOptions, UpdateOneModel, UpdateOptions, WriteModel}
+import com.mongodb.spark.MongoConnector
+import com.mongodb.spark.config.WriteConfig
+import com.mongodb.spark.sql.MapFunctions.rowToDocumentMapper
+import com.mongodb.spark.sql.v2.MongoDataSourceWriter.SaveType
+import org.apache.spark.sql.sources.v2.writer.{DataSourceWriter, DataWriter, DataWriterFactory, WriterCommitMessage}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{Row, SaveMode}
+import org.bson.BsonDocument
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+object MongoDataSourceWriter {
+  object SaveType extends Enumeration {
+    type SaveType = Value
+    val Insert, Update, Replace = Value
+  }
+}
+
+case class MongoDataSourceWriter(private val writeConfig: WriteConfig, private val schema: StructType, private val mode: SaveMode) extends DataSourceWriter {
+  private val mapper = rowToDocumentMapper(schema)
+  private val fieldNames = schema.fieldNames.toList
+  private val queryKeyList = BsonDocument.parse(writeConfig.shardKey.getOrElse("{_id: 1}")).keySet().asScala.toList
+  private val containsQueryKeyList = queryKeyList.forall(fieldNames.contains)
+  private val saveType = {
+    if (mode == SaveMode.Overwrite || writeConfig.forceInsert) {
+      SaveType.Insert
+    } else if (containsQueryKeyList) {
+      if (writeConfig.replaceDocument) SaveType.Replace else SaveType.Update
+    } else {
+      SaveType.Insert
+    }
+  }
+  val mongoConnector = MongoConnector(writeConfig.asOptions)
+
+  override def createWriterFactory(): DataWriterFactory[Row] = MongoDataWriteFactory()
+
+  override def commit(messages: Array[WriterCommitMessage]): Unit = {
+  }
+
+  override def abort(messages: Array[WriterCommitMessage]): Unit = {
```

Comment:
The Javadoc says 

>    * If this method fails (by throwing an exception), the underlying data source may require manual cleanup.

Is that not the case for MongoDB without transactions?

Reply:
This is an API that is going through further change.  Essentially we could potentially clean up written messages here.